lib/Assertion/ExpressionParser.php

```diff
@@ -13,188 +13,299 @@
 namespace PhpBench\Assertion;
 
 use PhpBench\Assertion\Ast\Comparison;
+use PhpBench\Assertion\Ast\FloatNode;
+use PhpBench\Assertion\Ast\FunctionNode;
+use PhpBench\Assertion\Ast\IntegerNode;
 use PhpBench\Assertion\Ast\MemoryValue;
 use PhpBench\Assertion\Ast\Node;
+use PhpBench\Assertion\Ast\NumberNode;
 use PhpBench\Assertion\Ast\PercentageValue;
 use PhpBench\Assertion\Ast\PropertyAccess;
-use PhpBench\Assertion\Ast\ThroughputValue;
 use PhpBench\Assertion\Ast\TimeValue;
+use PhpBench\Assertion\Ast\ToleranceNode;
 use PhpBench\Assertion\Ast\Value;
-use PhpBench\Assertion\Ast\ZeroValue;
 use PhpBench\Assertion\Exception\SyntaxError;
-use PhpBench\Util\MemoryUnit;
-use PhpBench\Util\TimeUnit;
+use RuntimeException;
 
 class ExpressionParser
 {
+    /**
+     * @var Node[]
+     */
+    private array $parts = [];
+
     /**
      * @var ExpressionLexer
      */
     private $lexer;
 
+    public function __construct(ExpressionLexer $lexer)
+    {
+        $this->lexer = $lexer;
+    }
+
     public function parse(string $expression): Node
     {
-        $this->lexer = new ExpressionLexer($expression);
+        $this->lexer->setInput($expression);
```

Comment:
can refactor this class to purposefully build a stream of Nodes which can be popped and shifted